[cocoa] Refine when opt-in partitioned cookies is enabled

sysrqb opened the pull request:
#### 9b3c87ad1e54bafd9b898d2af88bee03706a02d6
<pre>
[cocoa] Refine when opt-in partitioned cookies is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=297513">https://bugs.webkit.org/show_bug.cgi?id=297513</a>
<a href="https://rdar.apple.com/158511507">rdar://158511507</a>

Reviewed by Mike Wyrzykowski.

In 298679@main, we changed how the feature is enable so it is guarded by
CFN_COOKIE_ACCEPTS_POLICY_PARTITION. That doesn&apos;t work because the necessary
header isn&apos;t included. Therefore, this change takes a different approach. It
only references the guard within files where the macro is defined. In all other
places, we guard the implementation by HAVE(OPT_IN_PARTITIONED_COOKIES). This
required moving some of the compile-time guards. In some cases this resulted in
an empty implementation when the guard is false, or defining a partial
implementation. In a couple cases, this required adding platform-specific
implementations on Cocoa.

The result of this change is that enabling the web preference is necessary but
not sufficient for enabling the feature. The preference only guards an empty
implementation, by itself.

This patch introduces a new invariant that when the CFN macro is not defined,
third-party cookie blocking is never
ThirdPartyCookieBlockingMode::AllExceptPartitioned.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setOptInCookiePartitioningEnabled):
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::policyProperties):
(WebCore::NetworkStorageSession::httpCookiesForURL const):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::setCookieFromDOM const):
(WebCore::NetworkStorageSession::domCookiesForHost):
(WebCore::NetworkStorageSession::setOptInCookiePartitioningEnabled):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::createWebSocketTask):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::setCookieTransform):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
(WebKit::WebSocketTask::WebSocketTask):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::thirdPartyCookieBlockingMode const):
(WebKit::WebsiteDataStore::isOptInCookiePartitioningEnabled const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setThirdPartyCookieBlockingMode):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieCacheCocoa.mm:
(WebKit::WebCookieCache::inMemoryStorageSession):
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):

Canonical link: <a href="https://commits.webkit.org/298824@main">https://commits.webkit.org/298824@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77d7fd7a9654a4f662c7b5ca3fe3558274033a89

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27012 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67359 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6a9e8d5f-4cdf-4771-840f-53808295442c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37136 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45039 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88676 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43092 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/96009a9e-d118-49d1-b438-6bd6604b47b2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119723 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29608 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104755 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69146 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28670 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22860 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66515 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108888 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98987 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23016 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125984 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115302 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43673 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32796 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97345 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100957 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97142 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42468 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20408 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39660 "Hash 77d7fd7a for PR 49508 does not build (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18648 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43559 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49154 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144000 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43026 "Built successfully") | | [💥 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37073 "An unexpected error occured. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46365 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44731 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/77d7fd7a9654a4f662c7b5ca3fe3558274033a89. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49508

Commit:
[cocoa] Refine when opt-in partitioned cookies is enabled
https://bugs.webkit.org/show_bug.cgi?id=297513
rdar://158511507

Reviewed by Mike Wyrzykowski.

In 298679@main, we changed how the feature is enable so it is guarded by
CFN_COOKIE_ACCEPTS_POLICY_PARTITION. That doesn't work because the necessary
header isn't included. Therefore, this change takes a different approach. It
only references the guard within files where the macro is defined. In all other
places, we guard the implementation by HAVE(OPT_IN_PARTITIONED_COOKIES). This
required moving some of the compile-time guards. In some cases this resulted in
an empty implementation when the guard is false, or defining a partial
implementation. In a couple cases, this required adding platform-specific
implementations on Cocoa.

The result of this change is that enabling the web preference is necessary but
not sufficient for enabling the feature. The preference only guards an empty
implementation, by itself.

This patch introduces a new invariant that when the CFN macro is not defined,
third-party cookie blocking is never
ThirdPartyCookieBlockingMode::AllExceptPartitioned.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setOptInCookiePartitioningEnabled):
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::policyProperties):
(WebCore::NetworkStorageSession::httpCookiesForURL const):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::setCookieFromDOM const):
(WebCore::NetworkStorageSession::domCookiesForHost):
(WebCore::NetworkStorageSession::setOptInCookiePartitioningEnabled):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::createWebSocketTask):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::setCookieTransform):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
(WebKit::WebSocketTask::WebSocketTask):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::thirdPartyCookieBlockingMode const):
(WebKit::WebsiteDataStore::isOptInCookiePartitioningEnabled const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setThirdPartyCookieBlockingMode):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieCacheCocoa.mm:
(WebKit::WebCookieCache::inMemoryStorageSession):
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):

Canonical link: https://commits.webkit.org/298824@main
